src/basset.cpp

```diff
@@ -190,79 +160,107 @@ void Regex::report(int errcode) const {
   throw;
 }
 
-class CompilationDatabase : ofstream {
+class CompilationDatabase {
 public:
-  template <typename... Args> explicit CompilationDatabase(Args &&...);
+  using IsSourceFileFunc = function<bool(const string &)>;
+
+  CompilationDatabase(string filename, IsSourceFileFunc is_source_file);
+
   CompilationDatabase(const CompilationDatabase &) = delete;
   CompilationDatabase(CompilationDatabase &&) = delete;
   CompilationDatabase &operator=(const CompilationDatabase &) = delete;
   CompilationDatabase &operator=(CompilationDatabase &&) = delete;
-  ~CompilationDatabase() override;
-  using ofstream::operator!;
+  ~CompilationDatabase() = default;
+
+  bool load();
+  bool save();
+
   void add(const string &directory, const vector<string> &command);
 
 private:
-  [[nodiscard]] static bool is_source_file(const string &argument);
-  bool first{true};
+  const string filename;
+  const IsSourceFileFunc is_source_file;
+
+  static string make_index_key(const string &directory,
+                               const string &filename) {
+    return canonical(path(directory) / filename);
+  }
+
+  // maps canonized source file path to (directory, arguments) pair
+  unordered_map<string, pair<string, vector<string>>> index;
```

Comment:
Please consider using `path` as key type:
```suggestion
  unordered_map<path, pair<string, vector<string>>> index;
```